rig-core/src/one_or_many.rs

```diff
@@ -0,0 +1,278 @@
+/// Struct containing either a single item or a list of items of type T.
+/// If a single item is present, `first` will contain it and `rest` will be empty.
+/// If multiple items are present, `first` will contain the first item and `rest` will contain the rest.
+/// IMPORTANT: this struct cannot be created with an empty vector.
+/// OneOrMany objects can only be created using OneOrMany::from() or OneOrMany::try_from().
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct OneOrMany<T> {
+    /// First item in the list.
+    first: T,
+    /// Rest of the items in the list.
+    rest: Vec<T>,
+}
+
+/// Error type for when trying to create a OneOrMany object with an empty vector.
+#[derive(Debug)]
+pub struct EmptyListError;
+
+impl std::fmt::Display for EmptyListError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Cannot create OneOrMany with an empty vector.")
+    }
+}
+impl std::error::Error for EmptyListError {}
```

Comment:
Use `derive(thiserror::Error)` instead of implementing `Display` and `Error` manually